src/sdglib/factories/Kmers.h

```diff
@@ -0,0 +1,181 @@
+        }
+    protected:
+        const Kmers& parent;
+        const char* position;
+        value_type fwmer;
+        value_type rvmer;
+        value_type mask;
+        size_t last_unknown;
+
+        void scan_right() {
+            while(*position != '\0') {
+                ++position;
+                const char nucleotide = *position;
+                // Get bits for the nucleotide.
+                // The array should be a static and constant array.
+                incorporate_char(nucleotide);
+                if(last_unknown >= parent.K) break;
+            }
+        }
+
+        void incorporate_char(const char c) {
+            value_type fbits = parent.translation_table[c];
+            value_type rbits = ~fbits & (value_type) 3;
+            last_unknown = fbits == 4 ? 0 : last_unknown + 1;
+            fwmer = (fwmer << 2 | fbits);
+            rvmer = (rvmer >> 2) | (rbits << ((parent.K - 1) * 2));
+        }
+    };
+    // An iterator at the first valid Kmer in the sequence.
+    iterator begin() {
+        iterator it(*this);
+        ++it;
+        return it;
+    }
+    // An iterator past all valid Kmers in the sequence (has hit \0 in input string).
+    iterator end() {
+        iterator it = begin();
+        while(!it.has_reached_end()) ++it;
+        return it;
+    }
+private:
+    // std::string& sequence;
+    // Doesn't this make it less safe than a std::string reference though?
+    // Someone could just free the memory out from underneath us.
+    const char* sequence;
+    const int K;
+    std::array<U, 256> translation_table;
+};
+
+template<typename U>
+class CanonicalKmers : public Kmers<U> {
+    using value_type = typename Kmers<U>::value_type;
+    using reference = typename Kmers<U>::reference;
+    using pointer = typename Kmers<U>::pointer;
+    using Kmers<U>::Kmers; // Inherit constructor.
+
+public:
+    class iterator : public Kmers<U>::iterator {
+    public:
+        explicit iterator(CanonicalKmers& p) : Kmers<U>::iterator(p) {}
+        using Kmers<U>::iterator::operator++;
+        reference operator*() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return std::min(Kmers<U>::iterator::fwmer & Kmers<U>::iterator::mask, Kmers<U>::iterator::rvmer & Kmers<U>::iterator::mask);
+        }
+        pointer operator->() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return std::min(Kmers<U>::iterator::fwmer & Kmers<U>::iterator::mask, Kmers<U>::iterator::rvmer & Kmers<U>::iterator::mask);
+        }
+    };
+
+    iterator begin() {
+        iterator it(*this);
+        ++it;
+        return it;
+    }
+    // An iterator past all valid Kmers in the sequence (has hit \0 in input string).
+    iterator end() {
+        iterator it = begin();
+        while(!it.has_reached_end()) ++it;
+        return it;
+    }
+};
```

Comment:
I would implement this differently: a non-canonical fast version that only populates fwmer and where * gives you fwmer, and a canonical version where fwmer rvmer and canmer are actually in a struct as (*it).fkmer (*it).rkmer (*it).ckmer with this struct (class??) also giving access to the position in the sequence and an is_fw() method that returns ckmer==fkmer.
This will give us a bit of flexibility in implementations and does not sound too terribly onerous in computing if done right. Use this same struct/class to keep the internal values and the overhead will be small (ish)

Reply:
I've added a class that has fkmer, rkmer, ckmer & is_fwd etc methods.